src/components/modal/ModalProvider.tsx

```diff
@@ -2,30 +2,36 @@ import React from 'react';
 import { ModalLauncher } from './createModalLauncher';
 import { ModalPortal } from './ModalPortal';
 
-type ModalContextType = (launcher: ModalLauncher) => void;
+type ModalContextType<T = {}> = (launcher: ModalLauncher) => { closed?: Promise<T> };
 
-const ModalContext = React.createContext<ModalContextType>(() => {});
+const ModalContext = React.createContext(
+  // eslint-disable-next-line
+  <T extends {}>(launcher: ModalLauncher): { closed?: Promise<T> } => ({
+    closed: Promise.resolve<T>({} as T),
+  }),
+);
 
 type ModalProviderProps = {
   children: React.ReactNode;
 };
 
-export const useModalLauncher = () => React.useContext(ModalContext);
+export const useModalLauncher = <T extends {}>() =>
+  React.useContext<ModalContextType<T>>(ModalContext);
 
 export const ModalProvider: React.FC<ModalProviderProps> = ({ children }) => {
   const [modal, setModal] = React.useState<React.ReactElement>();
 
-  const handleClose = React.useCallback(() => {
-    setModal(null);
-  }, []);
-
-  const showModal = React.useCallback(
-    (launcherCallback) => {
+  const showModal = React.useCallback(<T extends {}>(launcherCallback): { closed?: Promise<T> } => {
```

Comment:
And if we extract the type from context above, we can use it for the callback too.